src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+    rehashStep(s);
+}
+
+/* Allocates a new table and initiates incremental rehashing if necessary.
+ * Returns 1 on resize (success), 0 on no resize (failure). If 0 is returned and
+ * 'malloc_failed' is provided, it is set to 1 if allocation failed. If
+ * 'malloc_failed' is not provided, an allocation failure triggers a panic. */
+static int resize(hashset *s, size_t min_capacity, int *malloc_failed) {
+    if (malloc_failed) *malloc_failed = 0;
+
+    /* Adjust minimum size. We don't resize to zero currently. */
+    if (min_capacity == 0) min_capacity = 1;
+
+    /* Size of new table. */
+    signed char exp = nextBucketExp(min_capacity);
+    size_t num_buckets = numBuckets(exp);
+    size_t new_capacity = num_buckets * ELEMENTS_PER_BUCKET;
+    if (new_capacity < min_capacity || num_buckets * sizeof(bucket) < num_buckets) {
+        /* Overflow */
+        return 0;
+    }
+
+    signed char old_exp = s->bucketExp[hashsetIsRehashing(s) ? 1 : 0];
+    size_t alloc_size = num_buckets * sizeof(bucket);
+    if (exp == old_exp) {
+        /* The only time we want to allow resize to the same size is when we
+         * have too many tombstones and need to rehash to improve probing
+         * performance. */
+        if (hashsetIsRehashing(s)) return 0;
+        size_t old_num_buckets = numBuckets(s->bucketExp[0]);
+        if (s->everfulls[0] < old_num_buckets / 2) return 0;
+        if (s->everfulls[0] != old_num_buckets && s->everfulls[0] < 10) return 0;
+    } else if (s->type->resizeAllowed) {
+        double fill_factor = (double)min_capacity / ((double)numBuckets(old_exp) * ELEMENTS_PER_BUCKET);
+        if (fill_factor * 100 < MAX_FILL_PERCENT_HARD && !s->type->resizeAllowed(alloc_size, fill_factor)) {
+            /* Resize callback says no. */
+            return 0;
+        }
+    }
+
+    /* We can't resize if rehashing is already ongoing. Fast-forward ongoing
+     * rehashing before we continue. */
+    while (hashsetIsRehashing(s)) {
+        rehashStep(s);
+    }
+
+    /* Allocate the new hash table. */
+    bucket *new_table;
+    if (malloc_failed) {
+        new_table = ztrycalloc(alloc_size);
+        if (new_table == NULL) {
+            *malloc_failed = 1;
+            return 0;
+        }
+    } else {
+        new_table = zcalloc(alloc_size);
+    }
+    s->bucketExp[1] = exp;
+    s->tables[1] = new_table;
+    s->used[1] = 0;
+    s->rehashIdx = 0;
+    if (s->type->rehashingStarted) s->type->rehashingStarted(s);
+
+    /* If the old table was empty, the rehashing is completed immediately. */
+    if (s->tables[0] == NULL || s->used[0] == 0) {
+        rehashingCompleted(s);
+    } else if (s->type->instant_rehashing) {
+        while (hashsetIsRehashing(s)) {
+            rehashStep(s);
+        }
+    }
+    return 1;
+}
+
+/* Probing is slow when there are too many tombstones. Resize to the same size
+ * to trigger rehashing and cleaning up tombstones. */
+static int cleanUpTombstonesIfNeeded(hashset *s) {
```

Comment:
```suggestion
static int cleanUpEverfullFlagIfNeeded(hashset *s) {
```
Both tombstone and everful flag is used, I have a slight preference for everfullFlag  but would like it to be consistent.

Reply:
I have an idea for a more optimal way to set this flag. We actually don't need to set it when a bucket becomes full. We can wait a little longer. We need to set it when we first insert an element that doesn't fit in the full bucket and needs probing on insert. If we do this, then a better name for the flag is 'probe'.

I guess I had this in mind when I added this function.

Reply:
Sure, that name makes sense too.

BTW, I also had some other small suggestions for improvements, but held them off until the first iteration is merged.

Reply:
`cleanUpProbingChainsIfNeeded` is a better name I think.